### PR TITLE
[vtctld / wrangler] Extract some reparent methods out to functions for shared use between wrangler and VtctldServer

### PR DESCRIPTION
**Pull request by ajm188**
## Description

This PR extracts some reparent code from methods on the `Wrangler` struct to top-level functions, and adds tests for them. I'm doing this in advance to allow some code sharing between the legacy and new vtctld grpc APIs, to make those diffs smaller and less risky.

**Important**: This PR also contains a bugfix in `FindValidEmergencyReparentCandidates` (formerly `(*Wrangler).findValidReparentCandidates`) that would cause a non-deterministic panic, depending on what order the
`replicationStatusMap` is first iterated over in. The panic happens as follows:

Assume you have 2 tablets in the status map. One tablet has a GTID-based relay log position, and the other has a non GTID-based relay log position.

If the GTID tablet comes first, then `isGtidBased` gets set to a pointer to true, and when we come to the second, non-GTID tablet, we fall through to the 3rd case in the switch (`case status.RelayPosition.IsZero()`).

If the non-GTID tablet's position is zero, then we return an error, and everything's fine. _However_, if it's non-zero _and_ non GTID-based, then we successfully make it through the first loop, and later on, we try to `FindErrantGTIDs`, because `*isGtidBased` is true. This is where the panic happens, because we give it a mix of `Mysql56GTIDSet` and non-`Mysql56GTIDSet` positions.
<hr>
The fix here is to iterate through the full status map, and not break out early. We track whether we've seen _any_ GTID-based or non-GTID based positions, and fail early accordingly.

## Related Issue(s)

- #7058 
- #7428 

## Checklist
- [ ] Should this PR be backported? **No**
- [x] Tests were added or are not required
- [x] Documentation was added or is not required

## Impacted Areas in Vitess
Components that this PR will affect:

- [ ]  Query Serving
- [ ]  VReplication
- [x]  Cluster Management
- [ ]  Build/CI
- [ ]  VTAdmin

### CODE REVIEW

**Comment by ajm188** [comment]
These test failures in vtworker/testlib look legit, and related to my logic change in `StopReplicationAndBuildStatusMaps`; I'll look into them today:

```
--- FAIL: TestEmergencyReparentShard (0.03s)
Error:     vtctl_pipe.go:100: I0203 03:03:58.720256 replication.go:170] getting replication position from cell1-0000000001
Error:     vtctl_pipe.go:100: I0203 03:03:58.720358 replication.go:170] getting replication position from cell2-0000000003
Error:     vtctl_pipe.go:100: I0203 03:03:58.720879 replication.go:170] getting replication position from cell1-0000000002
Error:     vtctl_pipe.go:100: I0203 03:03:58.721351 replication.go:170] getting replication position from cell1-0000000000
Error:     vtctl_pipe.go:100: W0203 03:03:58.727724 replication.go:194] failed to get replication status from cell1-0000000000: rpc error: code = Unknown desc = TabletManager.StopReplicationAndGetStatus on cell1-0000000000 error: before status failed: no replication status: before status failed: no replication status
    emergency_reparent_shard_test.go:174: 
        	Error Trace:	emergency_reparent_shard_test.go:174
        	Error:      	Received unexpected error:
        	            	rpc error: code = Unknown desc = we got a filled-in relay log position, but it's not of type Mysql56GTIDSet, even thought we've determined we need to use GTID based assesment
        	Test:       	TestEmergencyReparentShard
```

and

```
--- FAIL: TestEmergencyReparentShardMasterElectNotBest (0.02s)
    emergency_reparent_shard_test.go:280: 
        	Error Trace:	emergency_reparent_shard_test.go:280
        	Error:      	"we got a filled-in relay log position, but it's not of type Mysql56GTIDSet, even thought we've determined we need to use GTID based assesment" does not contain "is not fully caught up"
        	Test:       	TestEmergencyReparentShardMasterElectNotBest
```

**Comment by ajm188** [comment]
Okay I've changed the strategy for the bugfix to use two booleans, one to keep track if we've ever seen a GTID-based relay log, and one to keep track if we've ever seen a non GTID-based relay log. Separately, we put _potential_ errors onto a `FirstErrorRecorder` if we see any empty, non GTID-based relay log positions.

Then, we fail if:
* we've seen any tablets with both GTID-based and non GTID-based relay logs _or_ 
* we've seen a tablet with an empty relay log position as well as another tablet with GTID-based relay log positions

```
> go test -v -race ./go/vt/vtctl/reparentutil/... ./go/vt/wrangler/testlib/... 
ok      vitess.io/vitess/go/vt/vtctl/reparentutil       0.983s
ok      vitess.io/vitess/go/vt/wrangler/testlib 15.413s
```

File: go/vt/vtctl/reparentutil/replication.go [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+	"vitess.io/vitess/go/vt/vttablet/tmclient"
+
+	replicationdatapb "vitess.io/vitess/go/vt/proto/replicationdata"
+	"vitess.io/vitess/go/vt/proto/vtrpc"
+)
+
+// FindValidEmergencyReparentCandidates will find candidates for an emergency
+// reparent, and, if successful, return a mapping of those tablet aliases (as
+// raw strings) to their replication positions for later comparison.
+func FindValidEmergencyReparentCandidates(
```
Comment (ajm188) [inline go/vt/vtctl/reparentutil/replication.go:42]: I'd like to call particular attention to this function in review. It's the only function I deliberately changed the behavior of, when I managed to write a test case that sometimes passed, and sometimes panicked. It's detailed in the PR description, but the tl;dr is that is I believe a mixed set of GTID-based and non GTID-based tablets were not always handled safely, and could result in panics from `FindErrantGTIDs`, depending on the order of iteration the first time we scan through `statusMap` to set `*isGtidBased` in the [original implementation](https://github.com/vitessio/vitess/blob/e3fc16e08cc44d5d39dad53a9ca54a620ba2fb37/go/vt/wrangler/reparent.go#L839-L850).

cc @deepthi

**Review by ajm188 — COMMENTED** [review]
Okay, @deepthi this should be ready for review now.

File: go/vt/vtctl/reparentutil/replication.go [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+		logger.Infof("getting replication position from %v", alias)
+
+		_, stopReplicationStatus, err := tmc.StopReplicationAndGetStatus(groupCtx, tabletInfo.Tablet, replicationdatapb.StopReplicationMode_IOTHREADONLY)
+		switch err {
+		case mysql.ErrNotReplica:
+			var masterStatus *replicationdatapb.MasterStatus
+
+			masterStatus, err = tmc.DemoteMaster(groupCtx, tabletInfo.Tablet)
+			if err != nil {
+				msg := "replica %v think it's master but we failed to demote it"
```
Comment (deepthi) [inline go/vt/vtctl/reparentutil/replication.go:184]: nit: think -> thinks

File: go/vt/vtctl/reparentutil/replication.go [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+
+	for alias, status := range replicationStatusMap {
+		if _, ok := status.RelayLogPosition.GTIDSet.(mysql.Mysql56GTIDSet); ok {
+			isGTIDBased = true
+		} else {
+			isNonGTIDBased = true
+		}
+
+		if status.RelayLogPosition.IsZero() {
+			// Potentially bail. If any other tablet hits the non-default
```
Comment (deepthi) [inline go/vt/vtctl/reparentutil/replication.go:71]: It took me a minute to understand this, it isn't clear what non-default means. Can you rephrase? Something like this (or your own phrasing is fine too)
```
// Record an error that we will only return if we have a mix of GTID and non-GTID tablets
```

File: go/vt/vtctl/reparentutil/util.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+	}
+
+	var (
+		searcher = topotools.NewMaxReplicationPositionSearcher(tmc, logger, waitReplicasTimeout)
+		wg       sync.WaitGroup
+	)
+
+	for _, tablet := range tabletMap {
+		switch {
+		case primaryCell != "" && tablet.Alias.Cell != primaryCell:
```
Comment (deepthi) [inline go/vt/vtctl/reparentutil/util.go:71]: This bothered me at first, but I think we looked at it once before and said that by default PRS won't choose a tablet in another cell, but you can always provide it a chosen tablet which will be respected. So I think this is fine.

**Review by deepthi — COMMENTED** [review]
Just a couple of nits, otherwise LGTM

**Comment by deepthi** [comment]
cc @PrismaPhonic in case you wanna take a look.

File: go/vt/vtctl/reparentutil/replication.go (hunk 3 above)
Comment (deepthi) [reply go/vt/vtctl/reparentutil/replication.go:71]: I still don't think I understood it correctly, so it's up to you to write the correct comment :)

**Comment by PrismaPhonic** [comment]
I'm confused by what real world use case would cause us to have mixed gtidset types.

**Comment by ajm188** [comment]
> I'm confused by what real world use case would cause us to have mixed gtidset types.

We don't support that. The change in implementation here is to detect it correctly and return an error. Previously, there were certain cases where we wouldn't catch it, call `status.FindErrantGTIDs` with the mixed types, which would cause a panic.

File: go/vt/vtctl/reparentutil/replication.go (hunk 3 above)
Comment (ajm188) [reply go/vt/vtctl/reparentutil/replication.go:71]: Yep, sorry, this was written at the time when I was using a switch case with a default here, and I forgot to go back and update the comment. Will fix.

File: go/vt/vtctl/reparentutil/util.go (hunk 4 above)
Comment (ajm188) [reply go/vt/vtctl/reparentutil/util.go:71]: Yeah, this is a preservation of functionality. Maybe later we can make a vtctld startup flag to turn this off (I'd prefer to be able to PRS cross-cell without having to pick the tablet myself), but definitely far beyond the scope of this change.

**Review by deepthi — APPROVED** [review]
